docs_src/design/ucr/Record-and-Replay.md

```diff
@@ -0,0 +1,54 @@
+
+Other potential uses are:
+
+- “Replayable data” could be used to validate a new or updated EdgeX service behaves like the existing one (EdgeX certification). Or the data could be offered by component providers to give users an example of the type of data generated from one of their device services.
+- EdgeX Ready applicants could use a record feature to capture data from their device service and the committee could explore that in relation to their profile to make sure they seemed to know what they were doing.
+- Record and replay can be used to reproduce a reported bug.
+- Record and export could serve as a simple means to backup data in the event of data loss.
+
+### Target Users
+- Device Manufacturer
+- Device Owner
+- Device User
+- Device Maintainer
+- Software Developer
+- Software Deployer
+- Software Integrator
+
+### Description
+Target users have the need to be able to replay recorded EdgeX Event/Readings for functional, performance or reproducible testing. This UCR describes a new capability that allows user to first record Event/Readings from real devices in real-time and than be able to replay the Event/Readings as if it was in real-time. The static device profile and device definition files at time of capture will need to be available and loaded at the time the captured data is replayed. 
+
+### Existing solutions
+There are simulators for some devices (i.e. Modbus), but there isn't an general solution to reproduce real device data into EdgeX without the physical devices being present. These simulators also don't have a way to produce a specific set of results in a timeline as do physical devices.
+
+### Requirements
+- Record device Event/Reading data (at rates, volumes and details like timestamps) 
+- Record must capture all Device definitions for devices referenced in the captured Event/Readings
+- Record must capture all Device Profiles for devices referenced in the captured Event/Readings
+- Export captured data for use at a later time or to send to other users
+- Import recoded data that was previously exported
+- Replay must add any captured Device definitions and Device Profiles to the system prior to play back.
+- Replay captured data at captured, slower or faster speeds
+- Replay recorded data from multiple sources at the same time (this mimics more device services feeding EdgeX )
+- Have the ability to record data only from specified target devices or device profiles.
+
```

Comment:
Should there be a requirement that the captures be in plain text (eg, JSON, YAML, csv, something) so they can be examined and altered with an editor? We have sometimes found that useful, especially when applied to automated testing.

Reply:
Sure, added to requirements and went with JSON since that is what Event/Readings are normally in and APIs for Device Profile and Devices return and accept JSON. CBOR will not be used for exported binary Readings.